Task #419 - GitHub Actions workflows

lovro-bikic opened the pull request:
Task: [#419](https://app.productive.io/1-infinum/tasks/2044901)

#### Aim

Replace Semaphore with GitHub Actions. 

#### Solution

Add reusable workflows for building and deploying. Add workflows which will be copied to the repo and which use the reusable workflows. Separate `bin/build` into `bin/lint` and `bin/test`.

lovro-bikic commented on workflows/build.yml:
```diff
@@ -0,0 +1,77 @@
+          POSTGRES_PASSWORD: postgres
+        ports:
+          - 5432:5432
+        options: --name=postgres
```
This is a Postgres service which will be used to run tests. There is nothing that special here (I pretty much followed best practices for setting this up), except a _tiny_ difference, which will also explain why I split `bin/build` into `bin/lint` and `bin/test`.

[Typically](https://docs.github.com/en/actions/using-containerized-services/creating-postgresql-service-containers#running-jobs-in-containers), Postgres is set up with the following options:
```yml
# Set health checks to wait until postgres has started
options: >-
  --health-cmd pg_isready
  --health-interval 10s
  --health-timeout 5s
  --health-retries 5
``` 

When you start a Postgres docker image, Postgres service usually isn't available immediately. Therefore, if one would run `docker run postgres` and immediately after that run `bundle exec rspec`, it would most probably fail because even though the container is running, the Postgres service wouldn't be ready at the time of executing tests.

This problem is usually solved by supplying a health check command which is run every couple of seconds until Postgres becomes ready (the above snippet). The process looks approximately like this:
- Start the container
- Run the `pg_isready` command, the container replies that Postgres isn't ready
- Sleep X seconds (The GA workflow is essentially blocked for X seconds)
- Run the `pg_isready` command, the container replies that Postgres isn't ready
- Sleep Y seconds
- Run the `pg_isready` command, the container replies that Postgres isn't ready
- Sleep Z seconds
- Run the `pg_isready` command, the container is ready!
- Run tests and everything else

The problem with this approach is having to wait all that time because `sleep` is invoked and nothing happens in the meantime (You can see an example of a workflow run with this approach [here](https://github.com/infinum/rails-infinum-revisor/runs/4622326769?check_suite_focus=true), the step you're looking for is `Initialize containers`). It takes 26-27 seconds on average to get Postgres up and running with this approach.

However, I opted for a different approach which saves us some time (remember that GA workflow runs are billed per minute, so time optimizations can be cost-effective in the long run).

This is my approach. We actually don't need Postgres until the point in the workflow in which tests are run, and this happens at the very end. In between initializing the Postgres service and actually connecting to Postgres, a lot happens (Ruby is set up, gems are installed, Node is set up, JS packages are installed, secrets are pulled, linters are run).

So, instead of waiting on Postgres to be ready, I only start the container and immediately start executing everything else. The Postgres service is still being prepared in the background while everything else is happening (those two things happen in parallel).

When everything else that doesn't require Postgres executes, and when we reach the point of needing Postgres, I then run this command: `until docker exec postgres pg_isready; do sleep 1; done`. This will run the same health check command `pg_isready` as in the first approach, and will also sleep if the service isn't ready yet. However, because of all the Ruby and Node setup, secrets pulling and linters running took some time to execute, by the time they're all done, Postgres will already be ready. In practice, `sleep 1` will never even execute, and tests will automatically follow after this command.

An example workflow run with this approach is [here](https://github.com/infinum/rails-infinum-revisor/runs/4688854144?check_suite_focus=true). This way, Postgres start up on average takes less than 10 seconds.

The latter approach is ~16 seconds faster than the original approach. I hope this will save us some billable time.

However, the faster approach isn't possible if we don't split `bin/build` into two files:
- part that doesn't depend on Postgres -> `bin/lint`
- part that depends on Postgres -> `bin/test`

The `bin/lint` part of the workflow also gives us some "free" time to wait for Postgres to be ready. I think it also conceptually makes sense to have linting and testing as separate scripts.

lovro-bikic commented on workflows/deploy.yml:
```diff
@@ -0,0 +1,45 @@
+name: Deploy
```
This is a reusable workflow, same as the build one above.

lovro-bikic commented on workflows/deploy.yml:
```diff
@@ -0,0 +1,45 @@
+
+on:
+  workflow_call:
+    inputs:
+      environment:
+        required: true
+        type: string
+      branch:
+        required: true
+        type: string
+      deployers:
+        required: false
+        type: string
+        default: ''
+      bundle_app_config:
+        required: false
+        type: string
+        default: .bundle/ci-deploy
```
These parameters must be provided by the workflow which references this reusable workflow.

`environment` is the same environment used for Mina (e.g. `production`).

`branch` is the git branch which will be checked out.

`deployers` is a list of people (their GH usernames) who are allowed to trigger a deploy via the UI. It can look like this:
```yml
deployers: '@lovro-bikic @melcha'
```
Anyone not in this list won't be allowed to deploy manually from UI. Note: everyone with write access to the repo can run the workflow from the UI, however the workflow will be stopped if they're not in the list.

`bundle_app_config` was explained above.

uncoverd replied:
This comment sounds like great documentation, can we include it in the workflow for future reference🙏

lovro-bikic replied:
Added in https://github.com/infinum/default_rails_template/pull/93/commits/71cc1741ea5f1407554c0f9599d06431fe902089.

JureCindro replied:
Could we change the `deployers:` to accept an array which is common practice in yml or is there a special restriction in github actions that requires a string with space delimited username handles?

lovro-bikic replied:
Inputs support only booleans, numbers, and strings sadly ([docs](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#onworkflow_callinputsinput_idtype)).

lovro-bikic commented on workflows/deploy.yml:
```diff
@@ -0,0 +1,45 @@
+        required: false
+        type: string
+        default: .bundle/ci-deploy
+    secrets:
+      SSH_PRIVATE_KEY:
+        required: true
```
This is a private SSH key which must be provided in order for SSH to work with the server (the server will have the public key).

lovro-bikic commented on workflows/deploy.yml:
```diff
@@ -0,0 +1,45 @@
+    runs-on: ubuntu-20.04
+    env:
+      BUNDLE_APP_CONFIG: ${{ inputs.bundle_app_config }}
+    if: ${{ github.event_name == 'workflow_dispatch' && contains(inputs.deployers, format('@{0}', github.actor)) || github.event.workflow_run.conclusion == 'success' }}
```
This is the condition under which the workflow is run. There are two cases: `workflow_dispatch` and `workflow_run`.

`workflow_dispatch` means manual trigger. In this case we check who the allowed deployers are (`contains(inputs.deployers, format('@{0}', github.actor))`).

`workflow_run` happens when a Build workflow run has completed. In this case we check if that run was successful (`github.event.workflow_run.conclusion == 'success'`).

d4be4st commented on deploy-production.yml:
```diff
@@ -0,0 +1,15 @@
+name: Deploy production
+
+on:
+  workflow_dispatch:
```
I do not think that is what you wrote.
The branch selection here is where do you want the workflow to be used from, not which branch is going to be "deployed".

At least that is how I understood it.

d4be4st replied:
arent you missing the `workflow_run` here as you do in staging?

lovro-bikic replied:
> arent you missing the workflow_run here as you do in staging?

Nope, because deploying to production should only be done manually (by default). If someone wants to automatically deploy to production on push to master, they can add the `workflow_run` event.

lovro-bikic replied:
> The branch selection here is where do you want the workflow to be used from, not which branch is going to be "deployed".

From the [docs](https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#workflow_dispatch):
<img width="763" alt="image" src="https://user-images.githubusercontent.com/36794420/148034117-b65f38d7-4d6d-4c23-840c-37c9813fd283.png">

So if you select `staging` branch for example and then use `actions/checkout`, it will check out `staging`.

d4be4st replied:
> Nope, because deploying to production should only be done manually (by default). If someone wants to automatically deploy to production on push to master, they can add the workflow_run event.

Hm... this is something we all need to agree on as its a major shift from what used to be. (Or per project base)
I vote for an automatic deploy as most code ending in master comes from PR merges.

nikajukic replied:
I think it should be configurable per project. 

On Loop I don't have automatic deploys at all because the deploy takes around 15mins and I usually bundle a few PRs together and then deploy them. 

On most other projects I don't have automatic deploy to production.

nikajukic replied:
@lovro-bikic can you add `workflow_run` event and comment it out with instructions on when/how to use it?

cilim replied:
I was also under the impression that all of our projects have automatic production deploys. Guess we don't 🤷 😄 

I'm okay with changing this workflow for all, but it would be cool to make it optional, so project teams can set this up as they want.

lovro-bikic replied:
> can you add workflow_run event and comment it out with instructions on when/how to use it?

Added.

d4be4st commented on template.rb:
```diff
@@ -680,6 +666,13 @@ def run
 
+## Users allowed to manually trigger deploys
+staging_deployers = ask('Who can manually trigger a deploy to staging? (Example: @username1 @username2)', :green)
+gsub_file('.github/workflows/deploy-staging.yml', 'DEPLOY USERS GO HERE', staging_deployers)
```
do we want this?

I see the appeal, but it might just be a hassle in the end...

How many people are distraught because they can rebuild/redeploy things in semaphore?


Also its easy to change the .yml file for a user and just push, no? :)

lovro-bikic replied:
> Also its easy to change the .yml file for a user and just push, no?

Not if the master branch is protected (and we typically have that now), so even if someone wants to change the list they first have to get approval from someone with master access (e.g. lead dev).

But as far as I know, not everyone with write access is automatically allowed to deploy to production, that's why I added this.

d4be4st commented on workflows/build.yml:
```diff
@@ -0,0 +1,77 @@
+        required: false
+        type: string
+        default: .bundle/ci-build
+    secrets:
+      VAULT_ADDR:
+        required: true
+      VAULT_AUTH_METHOD:
+        required: true
+      VAULT_AUTH_USER_ID:
+        required: true
+      VAULT_AUTH_APP_ID:
+        required: true
```
here is a question to think about:

Do we maybe want to change this us? Istead of pulling from vault just add a signle secret as the contents of the whole application.yml file?

and have a `run: echo $SECRETS > config/application.yml` as a job?


Pros: 
We reduce the need for GHA to have access to our vault
One less user/app creds on vault for devops to create

Cons:
you need an admin to update the secrets

lovro-bikic replied:
I'm not sure, but the immediate problem that I see is having to rely on two sources of truth for secrets. One source of truth is Vault, the other is GitHub. If you update one but not the other, you'll run into problems (and people are forgetful, I would most probably forget to update secrets on GH). I think it's better that we have a single source of truth.

d4be4st commented on workflows/build.yml:
```diff
@@ -0,0 +1,77 @@
+      - uses: ruby/setup-ruby@v1
+        with:
+          bundler-cache: true
+      - uses: actions/setup-node@v2
```
so we need 3 steps for node while ruby has that all in one step? 

🤦

lovro-bikic replied:
Yes, because `setup-node` actions doesn't run `yarn install` (`setup-ruby` on the other hand runs `bundle install`), and I need to cache node modules (`setup-node` can easily miss cache so I had to set it up manually (see [here](https://github.com/actions/setup-node/issues/328))).

d4be4st commented on workflows/build.yml:
```diff
@@ -0,0 +1,77 @@
+      - name: Install JS packages
+        if: ${{ inputs.use_node }}
+        run: yarn install
+      - name: Run linters
```
in one of my previous ventures i setup the linters to run paralel to the specs
so each linter is run independently and once all have been run then the specs are run

it might save a couple of real world minutes because its parallel :)

uncoverd replied:
![Screenshot 2022-01-10 at 11 03 48](https://user-images.githubusercontent.com/7644519/148748169-f07a4637-229d-405f-8bc8-f9081b1ae2f6.png)

Agreed, I think we care much more about the real world minutes then the billed minutes, our hour is not cheap 😄

lovro-bikic replied:
I can separate the workflow into separate jobs, but I have some concerns.

First, pricing. GHA charges each job separately. In the above example, `bundle-cache` is charged 1 billable minute, `specs` is charged 4 billable minutes, `lint` is charged 2 billable minutes, and `audit` is charged 1 billable minute. The total is 8 billable minutes.

Each of those jobs invokes the checkout action, and setup-ruby action. This typically takes 10 seconds (if we assume that no new gems are downloaded, i.e. everything is pulled from cache).

Compare this to a single job which does the checkout, Ruby setup, audit, specs, and linters. With the above numbers, it would approximately take:
10s (checkout, setup) + 2m 50s (specs - checkout, setup) + 1m 20s (lint - checkout, setup) + 10s (audit - checkout, setup) = 4m 30s

This is 5 billable minutes. The time difference is 3 billable minutes, and the cost difference is $0,024 (Linux costs $0,008 per billable minute). In itself, that's not costly. However, these jobs will be run hundreds of times on tens of projects. Suddenly, 3 billable minutes become 3000 billable minutes. And costs aggregate with time, so at the end of the year, the cost difference becomes even greater.

I generally don't have a problem with GHA costing more if that's appropriate for our use-case, but I just want everyone to be on the same page about the cost aspect.

My second concern is, why would we want to run specs if linting and/or audit fails? If any one of those jobs fails, the workflow run is considered failed.

JureCindro replied:
So we can compare the build times, but include the price of the developer in it. I think it is a safe assumption that during the build process, the developer is not doing anything other than waiting, on average. 
I know some (hyper-productive) people want to say that they reply on slack, reply on emails, etc during the build process. But often those tasks take more than the build time so then the merge is waiting. Maybe even someone else merges while the busy-body developer is replying to his mails and then it requires him to rebase the branch before merging again. I for one am so small minded that if I do anything else (productive at least) in between, I completely forget about the branch build.

JureCindro replied:
So we can compare the build times, but include the price of the developer in it. I think it is a safe assumption that during the build process, the developer is not doing anything other than waiting, on average. 
I know some (hyper-productive) people want to say that they reply on slack, reply on emails, etc during the build process. But often those tasks take more than the build time so then the merge is waiting. Maybe even someone else merges while the busy-body developer is replying to his mails and then it requires him to rebase the branch before merging again. I for one am so small minded that if I do anything else (productive at least) in between, I completely forget about the branch build, and then merge half a day later when someone pings me regarding this.

So for sake of the exercise let's assume that the developer is waiting during build process that is required for the merge. I believe the billed hour is a bit more than 50 eur/h.
As you stated, the difference is 3 billable minutes. This saves $0,024 in container running time. The difference in developer billable minutes is thus `50 / 60 * 3` which is 2,5 €.
So in the example where we parallelise the build runs we save more than 100x the amount we pay more for the billable time of the runners.

For the second concern, I think it is beneficial to know that specs have completed successfully but only the audit failed. Same for any other combination. You'd also want to know if all of those fail, and not exit early if any of these 3 fail.

d4be4st replied:
To save time, can we introduce tags to run just specific jobs?
For example, rubocop linter failed because you had more than 120 line, and you want to fix it but you really just want to run the linter again

also [skip ci] tag would be handy to skip everything.

maybe something for the next iteration?

melcha replied:
@lovro-bikic could checkout and ruby setup be extracted into a separate job, and then parallel lint, test and audit jobs would depend on the extracted setup job? 
Maybe we could save some minutes this way.
https://docs.github.com/en/actions/learn-github-actions/managing-complex-workflows#creating-dependent-jobs

lovro-bikic replied:
> To save time, can we introduce tags to run just specific jobs?

Yes, this is possible. We can add this to the reusable workflow later on, and since it's backward-compatible, it can be added to v1.

cilim replied:
To save time, maybe it would be good to run the lint and audit checks **only** when a branch is merged to the staging/master branch? 🤔

d4be4st replied:
@cilim how are you going to "fix" the lint errors on master (protected branch) :D

melcha replied:
In a separate PR? It shouldn't happen that often because one should have rubocop embedded into their text editor and see rubocop offenses when editing files, and run overcommit locally before each commit like we did until now. 
I honestly still think only zeitwerk out of these check must be a hard check on CI, rubocop and bundle audit should be ignorable.

ivantomica replied:
Perhaps I'm going slightly of rails here, but my opinion was, and still is, we should not run Linting on each commit. It prolongs the build time with questionable usefulness and has the potential to cost us a lot.

What I would propose as an approach is:
- Linting - run locally using pre-commit/overcommit/whatever - also run on PR (I believe @cilim once mentioned that he first opens up PR and then starts developing in it - in my opinion - that's the wrong approach. If ABSOLUTELY necessary, alternative could be to explore possibility of ignoring Linting only on "Draft PR" feature Github has. And utilize that from the developer's end
- Security audit - should be part of the CI job on each PR to main branches - perhaps with option to merge regardless for the admins/TLs when situation requires it (:hankey: hits the fan and we need to fix something that's impairing production)
- Other checks - depends on what do they check :-) But we need to be sensible and ask ourselves "where do we get most value from this, while keeping cost in mind"

cilim replied:
> I believe @cilim once mentioned that he first opens up PR and then starts developing in it 

fyi: no 😄 I mostly develop everything before even pushing to remote

uncoverd replied:
I don't think we should be trading pennies for worse development outcomes. 

The only positive I see with reducing linting and bundle audit is that we save a few bucks. As @JureCindro said above, our development time is much more valuable than an external service's minute. In reality the test execution time takes the vast majority of the build time. The split we're seeing here is due to the team under @JureCindro watchful eye and their diligence to produce a well balanced test suite that has 100% coverage and runs extremely fast for the number of cases. Most other projects with at least decent coverage are much slower. In those cases linting and auditing becomes an even smaller portion of the bill.

Very few projects have been able to keep up with bundle audit and rubocop errors back in the day when these weren't mandatory on CI. I remember the old dance too well 

> "hey X, did you just push a change to staging that broke rubocop? God I hate you, I have to fix it in my staging merge now!
> "hey guys, did anyone notice that theres 5 bundle audit warnings?" "yeah I just ignored those, I needed to get 5 urgent fixes out"

@ivantomica I don't understand your argument for only running certain actions "on each PR", could you elaborate?
In general the CI pipeline runs when

- you push a branch for the first time (usually minutes before opening a PR)
- you push another commit to an existing branch (usually for PR fixes)
- you rebase the existing branch to the latest master (usually before merging)
- you merge a branch to master/uat/staging

I think we want the safety of the CI pipeline in all of these cases.
- When you first push a branch you want to see if its ready to be merged (auditing, linting, specs).
- Once you push a PR fix you again want to know if the change is safe (auditing, linting, specs).
- When you're just about ready to merge the feature branch to master you again want to be sure that the rebased version is still valid since its likely there were some code changes.
- Finally when you merge a feature branch into master you are constructing a new variation of the code that contains the original master + your feature branch changes and you again want to be sure the build passes as that code will get shipped.

I won't get too deep into the CI vs local debate again (we did this before, reference: https://github.com/infinum/rails-jsonapi-example-app/pull/42, https://github.com/infinum/default_rails_template/pull/63). All of these checks being done on each change automatically mean that the team fixes security issues earlier (since they break the build), stays in sync as far as code style is concerned and is confident that the code that is going out is always up to standards. This is in my opinion the cheapest win you can get, education and oversight is much more expensive.

@d4be4st I belive skip-ci already works out of the box, my gwip commits skip CI
![Screenshot 2022-01-12 at 19 17 15](https://user-images.githubusercontent.com/7644519/149198658-515f603e-2464-473d-a48f-27e638562def.png)

uncoverd replied:
I forgot one more thing - @lovro-bikic mentioned today that rubocop is awfully slow on CI. Locally it takes ~7s while on GHA it takes around a minute usually. 

I was able to halve this using the --parallel flag and I'm betting theres more execution speed to be gained via some config options or profiling.
![Screenshot 2022-01-12 at 20 14 19](https://user-images.githubusercontent.com/7644519/149206626-df667d0f-6383-417f-a556-a74498222d72.png)

lovro-bikic replied:
To summarize this whole thread: we've decided to run a single job, but run steps in parallel (using GNU parallel). This decision is a compromise: it's cheaper than the parallel jobs approach, but it still allows for auditing and linting on CI (for people who want that), without slowing down the single job.

Technically, this is achieved by splitting auditing, linting, and testing into three scripts, and calling them in parallel like this:
```
$ parallel --lb -k -j0 ::: bin/audit bin/lint bin/test
```
Flag explanations:
`--lb`: `line-buffer`, prints output line by line as it's produced (instead of waiting for the process to end before printing everything)
`-k`: `keep`, tells parallel not to mix output of different scripts (i.e. it will first output all lines from bin/audit, then all lines from bin/lint, and finally all lines from bin/test)
`-j0`: `jobs`, tells parallel to run as many jobs as possible in parallel. This will determined by the number of scripts parallel needs to run (by default it's 3, but some projects can use a different number of scripts)

d4be4st reviewed (commented):
Have you tested this out on any project?
I am wondering how the slacks /github integration has progressed.

Previously I had to use some slack/notifier actions to report to the slack which was not the best to be honest :D

lovro-bikic commented:
> Have you tested this out on any project?

@d4be4st it's currently live on [Revisor](https://github.com/infinum/rails-infinum-revisor/actions).

nikajukic commented:
@melcha @cilim @JureCindro @uncoverd @vr4b4c please take a look at this PR

nikajukic commented on deploy-staging.yml:
```diff
@@ -0,0 +1,19 @@
+    with:
+      environment: staging
+      branch: staging
+      deployers: 'DEPLOY USERS GO HERE'
```
Maybe write the example in the comment for what is expected here:
`'@github_username_1 @github_username_2'`

Same for deploy-production.yml

cilim commented on build.yml:
```diff
@@ -0,0 +1,16 @@
+    name: Build
+    uses: infinum/default_rails_template/workflows/build.yml@v1
+    with:
+      postgres_image: '14.1'
```
Is the postgres port by default 5432:5432? In case it's not, maybe write it as:

```suggestion
      postgres:
        image: postgres:14.1
```

Also, don't we need to setup some environment variables for this? For example, the postgres password or database name?

lovro-bikic replied:
> We don't have postgres 14.1 yet I think. Please check with devops. I think the latest RDS version we have is 13.2

Yep, RDS still doesn't offer version 14. I changed it back to 13.2.

cilim commented on workflows/build.yml:
```diff
@@ -0,0 +1,77 @@
+      postgres:
+        image: postgres:${{ inputs.postgres_image }}
+        env:
+          POSTGRES_USER: postgres
```
ah, this is setup here, got it! 👌

lovro-bikic commented on README.md:
```diff
@@ -23,6 +23,18 @@ then run if needed:
+- `VAULT_AUTH_USER_ID`
+- `VAULT_AUTH_APP_ID`
+
+For deploy workflows, you need to generate private/public SSH key pairs for each environment. Public key should be added to the server to which you're deploying. Private key should be added as a secret to GitHub and named `ENVIRONMENT_KEY`, where `ENVIRONMENT` is replaced with an appropriate environment name (`STAGING`, `PRODUCTION`, etc.).
```
Changed.

lovro-bikic replied:
@uncoverd This name was chosen in coordination with the Devops ([Slack ref](https://infinum.slack.com/archives/C0AME3E4Q/p1641294254126600)).

ivantomica replied:
When you phrase it like that, yes :-D Perhaps we need to change this on the WP side of things as well, to be on the same page.

As a favor towards DevOps team, perhaps you could say during setup process "we expect this to have ENVIRONMENT_SSH_KEY" or "ENVIRONMENT_DEPLOY_KEY" in environment variables on GHA"

ivantomica replied:
Checked with WP team, they're fine with the change. My hunch is telling me to go with ENVIRONMENT_DEPLOY_KEY. What do you think? Have preference one way or the other?

lovro-bikic replied:
I'd go with `SSH_PRIVATE_KEY_#{ENVIRONMENT}`, it's the most accurate description of the secret.

uncoverd commented on workflows/build.yml:
```diff
@@ -0,0 +1,77 @@
+          VAULT_AUTH_APP_ID: ${{ secrets.VAULT_AUTH_APP_ID }}
+      - name: Install JS packages
+        if: ${{ inputs.use_node }}
+        run: yarn install
```
I'm not sure exactly how it is with yarn, do we have to be careful here like we are with `npm install` vs `npm ci` ?
Ref: https://docs.npmjs.com/cli/v8/commands/npm-ci

lovro-bikic replied:
Good call, this behavior can be achieved in yarn with the `--frozen-lockfile` flag which I've now added.

uncoverd commented on workflows/deploy.yml:
```diff
@@ -0,0 +1,45 @@
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
```
Should we really cancel in progress deploys here? That might get us into weird deployment states which could be tricky to recover from 🤔

lovro-bikic replied:
True, I removed `cancel-in-progress`. I left the `group` part of the concurrency setup, so this will at least cancel pending jobs if there is more than one of them.

lovro-bikic commented on workflows/build.yml:
```diff
@@ -0,0 +1,77 @@
+jobs:
+  lint_and_test:
+    name: 'Lint and test'
+    runs-on: ubuntu-20.04
```
I can add a variable for this (with the default set to `ubuntu-20.04`), but what other runner would you put here? There's Windows, Mac, and an older version of Ubuntu.

uncoverd replied:
22.04 in April 🤞 

![Screenshot 2022-01-10 at 12 12 33](https://user-images.githubusercontent.com/7644519/148756963-fb525802-2275-41a1-b2a5-1c68fa961728.png)

lovro-bikic replied:
I added an optional input for a runner with the default set to `ubuntu-20.04`.

Just a note, the idea with reusable workflows is that at one point when a new runner becomes available (e.g. Ubuntu 22.04), we'll only have to update the runner in the reusable workflow, and all projects will be using Ubuntu 22 as well immediately (if it's not a breaking change, of course).

vr4b4c commented:
👏

cilim reviewed (approved):
Awesome stuff! 👏

melcha commented on template.rb:
```diff
@@ -228,6 +210,10 @@
 create_file 'bin/deploy', BIN_DEPLOY, force: true
```
We might need to set `RAILS_ENV=test` for `db:test:prepare` explicitly.

Remember my PR from last week, I removed export RAILS_ENV=test from bin/deploy? I tested that on Cormeum, and the Semaphore deploy failed at this line because:

1. the default RAILS_ENV is development, if nothing is set
2. annotate gem is in development group in gemfile
3. we skip development group in `.bundle/ci-deploy/config`
4. lib/tasks/auto_annotate_models.rake checks if the rails env is development and then requires the gem and defines the task.

So `RAILS_ENV=test bundle exec rails db:test:prepare` works.

lovro-bikic replied:
I added it back.

melcha reviewed (approved):
Excellent, @lovro-bikic !

lovro-bikic commented on template.rb:
```diff
@@ -228,6 +210,10 @@
 create_file 'bin/deploy', BIN_DEPLOY, force: true
 chmod 'bin/deploy', 0755, verbose: false
 
+get("#{BASE_URL}/build.yml", '.github/workflows/build.yml')
```
This is at the top of the template:
```ruby
BASE_URL = 'https://raw.githubusercontent.com/infinum/default_rails_template/master'.freeze
```

lovro-bikic commented on template.rb:
```diff
@@ -670,6 +656,8 @@ def run
 
   run 'yarn add --dev @infinumrails/eslint-config-js @infinumrails/stylelint-config-scss eslint postcss stylelint'
+
+  gsub_file('.github/workflows/build.yml', /^.*use_node: false.*\n/, '')
```
So by default Node is turned off in the build workflow. This line is part of the conditional where the user is asked "Will this application have a frontend?", and if they respond yes, then Node installation is enabled by removing this line.

lovro-bikic commented on build.yml:
```diff
@@ -0,0 +1,16 @@
+jobs:
+  build:
+    name: Build
+    uses: infinum/default_rails_template/workflows/build.yml@v1
```
@melcha can you set that up? I pushed the branch (it's currently on the same commit as master).

melcha replied:
Done.

uncoverd reviewed (approved):
🙆‍♂️

JureCindro commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,108 @@
+        required: true
+      VAULT_AUTH_APP_ID:
+        required: true
+
+jobs:
+  build:
```
Should we add some default env variables here?

```yml
env:
  RAILS_ENV: test
  BUNDLE_WITHOUT: development
  ```

lovro-bikic replied:
`BUNDLE_WITHOUT` is set through the `BUNDLE_APP_CONFIG` variable, which is defined per-project.

`RAILS_ENV` has been added where it's necessary (`bin/test`), but I don't see any harm in adding in globally.

lovro-bikic commented:
There have been some changes since the last review, please take a look at the PR again.

Important changes:
- added `bin/audit` which has some of the steps from `bin/lint` (https://github.com/infinum/default_rails_template/pull/93/commits/ab144e3086fd0e97d1b53501a328c7ea7aa6ae26)
- added `bin/prepare_ci` instead of the hardcoded step for pulling secrets. I noticed that some projects have a more elaborate CI setup, so this gives greater flexibility to those projects (https://github.com/infinum/default_rails_template/pull/93/commits/5c2a87553b715ed11bfcb9d7cda2b97a9a7e6d73)
- steps were changed to run in parallel with the [parallel](https://www.gnu.org/software/parallel/) command. additionally, an input `ci_steps` was added so projects can define which files they want to run (by default, it's `bin/audit bin/lint bin/test`, but if a project has a different setup, they can override it).  (https://github.com/infinum/default_rails_template/pull/93/commits/5c2a87553b715ed11bfcb9d7cda2b97a9a7e6d73) (more info [here](https://infinum.slack.com/archives/C027RRR17/p1641996025020100))
- Added RuboCop cache. On one project, the cache reduced RuboCop execution time from 60s to 1s (https://github.com/infinum/default_rails_template/pull/93/commits/14869808ddb86b05ea5ab049a7ed9c3bfb7aa48a)

lovro-bikic commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,107 @@
+
+      # Defines which scripts will run on CI
+      # For the format, see the default value
+      ci_steps:
+        required: false
+        type: string
+        default: 'bin/audit bin/lint bin/test'
```
I wasn't sure whether it's better to set a default value here, or make the input required so that each project has to define in their own build workflow which steps to use. The second approach is more explicit, and developers can look into their project build.yml to figure out which scripts will run on CI. What do you think?

lovro-bikic replied:
After a bit of thinking, I decided to make this input required and defined per-project. That way, developers can know how their CI will work just by looking at their `build.yml`. (https://github.com/infinum/default_rails_template/pull/93/commits/633753171a074db7d6c911d1417db44f857788d8)

lovro-bikic commented on template.rb:
```diff
@@ -135,55 +135,67 @@
-
   echo "=========== rubocop  ==========="
-  time bundle exec rubocop --format simple
+  time bundle exec rubocop --format github --color --parallel
```
This is how it looks in the Files changed tab:
<img width="1249" alt="image" src="https://user-images.githubusercontent.com/36794420/149338365-1a9c5919-37fe-4f35-9b7c-8ae3f65f91ee.png">

uncoverd reviewed (approved):
🔥
